src/main/java/org/wildfly/security/auth/client/MatchHostRule.java

```diff
@@ -25,30 +25,48 @@
 import java.util.Locale;
 import java.util.regex.Pattern;
 
+import org.wildfly.common.net.Inet;
+
 /**
  * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
  */
 class MatchHostRule extends MatchRule {
 
-    private static final Pattern validHostSpecPattern = Pattern.compile("(?i:[-a-z0-9_]+(?:\\.[-a-z0-9_]+)*\\.?)");
+    private static final Pattern validHostSpecPattern = Pattern.compile("(?i:[-a-z0-9_]+(?:\\.[-a-z0-9_]+)*)");
 
     private final String hostSpec;
 
-    MatchHostRule(final MatchRule parent, final String hostSpec) {
+    MatchHostRule(final MatchRule parent, String hostSpec) {
         super(parent);
-        if (! validHostSpecPattern.matcher(hostSpec).matches()) {
-            throw log.invalidHostSpec(hostSpec);
-        }
-        if (hostSpec.endsWith(".")) {
-            this.hostSpec = hostSpec.substring(0, hostSpec.length() - 1).toLowerCase(Locale.ENGLISH);
-        } else {
-            this.hostSpec = hostSpec.toLowerCase(Locale.ENGLISH);
+        if (hostSpec.contains(":")) { // IPv6
+            byte[] bytes = Inet.parseInet6AddressToBytes(hostSpec);
+            if (bytes == null) {
+                throw log.invalidHostSpec(hostSpec);
+            }
+            this.hostSpec = Inet.toOptimalString(bytes);
+        } else { // IPv4 or domain name
+            if (!validHostSpecPattern.matcher(hostSpec).matches()) {
+                throw log.invalidHostSpec(hostSpec);
+            }
+            this.hostSpec = hostSpec.toLowerCase(Locale.ROOT);
         }
     }
 
     public boolean matches(final URI uri, final String abstractType, final String abstractTypeAuthority, final String purpose) {
         String host = uri.getHost();
-        return host != null && host.startsWith(hostSpec) && super.matches(uri, abstractType, abstractTypeAuthority, purpose);
+        if (host == null) {
+            return false;
+        }
+
+        if (host.startsWith("[") && host.endsWith("]")) {
+            // normalize IPv6 address
+            byte[] bytes = Inet.parseInet6AddressToBytes(host);
+            if (bytes != null) {
+                host = Inet.toOptimalString(bytes);
+            }
+        }
+
+        return host.toLowerCase(Locale.ROOT).equals(hostSpec) && super.matches(uri, abstractType, abstractTypeAuthority, purpose);
```

Comment:
I think we want to match by bytes not by string though, if an IP address is given.  So if the host is an IP address and the match criterion is an IP address, then there's a match if the bytes are equal; if the host is a string and the match criterion is a string, then there's a match if the lowercased strings are equal; otherwise there's no match.